doc/man/gnunet-pseudonym.1
.TH GNUNET-PSEUDONYM "1" "04 Nov 2004" "GNUnet"
.SH NAME
gnunet\-pseudonym \- create, delete or list pseudonyms
.SH SYNOPSIS
.B gnunet\-pseudonym 
[options]
.SH DESCRIPTION
.PP
gnunet\-pseudonym is a tool for managing pseudonyms and namespaces.  A pseudonym is the persona that controls a namespace.  As such, it is identical to a public\-private RSA key pair.  A namespace is a collection of files that have been signed by the corresponding private RSA key.  A namespace is typically associated with a nickname and other metadata which is kept in a specially named file in the namespace.  

Namespaces are an important tool for providing assurances about content integrity and authenticity in GNUnet.  Since all of the content in the namespace must have been provided by the same entity, users can form an opinion about that entity and learn to search (or avoid) certain namespaces.

gnunet\-pseudonym can be used to list all of the pseudonyms that were created locally, to create new pseudonyms, to delete existing pseudonyms (the namespace will continue to exist, but it will be impossible to add additional data to it) and to list all of the namespaces (with their meta-data) known to the local user.  By default, gnunet\-pseudonym lists all pseudonyms created locally and all of the namespaces that were discovered so far.  

Creating a new pseudonym requires using the \-C option together with a nickname that is to be used for the namespace.  Nicknames must be unique for each user, global uniqueness is desireable but not necessary.  If two namespaces in GNUnet use the same nickname all GNUnet tools will display the nickname together with the unique namespace identifier (which is derived from the public key and hence guaranteed to be unique) to avoid ambiguity.  Additional options can be passed together with the \-C option to provide additional meta-data that describes the namespace.  Possible meta-data includes the 'realname' of the person controlling the namespace, a description, the mime-type for content in the namespace (useful if the namespace is dedicated to some specific type of content) and contact information.  One important piece of meta-data that can be specified is the identifier of a document root, that is the name of a file in the namespace that is a portal to the rest of the content.  This is useful to help users find this root in the absence of conventions.  Note that all of this meta-data is optional and should never be trusted blindly.

When a pseudonym is created it can be optionally encrypted with a password.  That password is then required to insert content into the namespace and to determine that the pseudonym is available to the local user.  This option is useful if the local machine could fall under the control of an adversary and additional caution to protect the namespace is advisable.

gnunet\-pseudonym also lists the meta-data available for other namespaces.  Namespaces can be discovered whenever the peer obtains the namespace advertisement that is created at the time where the pseudonym is created.  Namespace advertisements can be found in directories (not implemented), ordinary keyword-based searches (by default gnunet\-pseudonym publishes the namespace advertisement under the keyword 'namespace', but the \-k option can be used to specify another keyword) and under the 'all-zeros' identifier of the respective namespace (using a namespace-search if the namespace ID is already known).

For more details about GNUnet namespaces and content encoding please read the 'Encoding for Censorship-resistant Sharing' (ECRS) paper which can be found on the GNUnet webpage.

.TP
\fB\-a\fR, \fB\-\-automate\fR
Start a (new) collection.  Works only in conjunction with the \-C option.  A collection is an automatically managed directory in a namespace.  In essence, after starting the collection every file that you insert into GNUnet will automatically be placed into the collection.  Other users can browse your collection and be certain (thanks to cryptography) that all of these files were inserted into GNUnet by the same user (they do not necessarily know who it is, but if you specify your realname (\-r) they will be able to see that handle).  Collections are useful for establishing a reputation for your GNUnet content, such that readers can form an opinion about quality and availability.  Namespaces can be used to achieve the same thing, but collections are automatic and thus less work for you.

You cannot use passwords to protect the collection.  Using collections has some security implications since it is possible for an adversary to see that all of these files originate from the same user.  This may help a correlation attack to break anonymity.  Nevertheless we encourage using collections, they are likely to be the right choice for most users. 
.TP
\fB\-C NAME\fR, \fB\-\-create=NAME\fR
Creates a new pseudonym with the given NAME.
.TP
\fB\-D NAME\fR, \fB\-\-delete=NAME\fR
Delete the pseudonym with the given NAME.
.TP
\fB\-e EMAIL\fR, \fB\-\-email=EMAIL\fR
Include EMAIL a contact address to contact the author of the namespace (use with \-C).
.TP
\fB\-E\fR, \fB\-\-end\fR
End a collection.  This option is the opposite of the \-a option in that it stops the collection.  Note that currently, once you stop a collection you can never restart it.  However, you can start a new collection.  There can only be one collection at any given point in time for a particular user.
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-k KEYWORD\fR, \fB\-\-keyword=KEYWORD\fR
Publish the namespace advertisement under the keyword 'KEYWORD'.  Default is 'namespace' (use with \-C).
.TP
\fB\-m MIMETYPE\fR, \fB\-\-mimetype=MIMETYPE\fR
Advertise that the namespace contains files of the given MIMETYPE (use with \-C).
.TP
\fB\-n\fR, \fB\-\-no\-advertisement\fR
Do not generate an advertisement for the namespace (use with \-C).
.TP
\fB\-p PASS\fR, \fB\-\-password=PASS\fR
Specify password to decrypt pseudonyms encountered or to encrypt pseudonyms created.  The namespace identifiers are displayed for the pseudonyms that could be decrypted.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Do not print the list of pseudonyms (only perform create or delete operation).
.TP
\fB\-r NAME\fR, \fB\-\-realname=NAME\fR
Claim that the name of the author of the content in the namespace in 'real' life is NAME (use with \-C).
.TP
\fB\-s ID:VALUE\fR, \fB\-\-set-rating=ID:VALUE\fR
Change the rating for the namespace identified by ID by VALUE.  For example, "\-s test:-3" decrements the rating of the pseudonym "test" by 3.  Note that ratings are purely local.  Each user has his own independent rating of namespaces.  The rating is merely a way for each user to keep track of his own experience with a given namespace.
.TP
\fB\-t TEXT\fR, \fB\-\-text=TEXT\fR
Use TEXT as the description for the namespace (use with \-C).
.TP
\fB\-u URI\fR, \fB\-\-uri=URI\fR
Include URI as an address where additional information about the namespace can be found (use with \-C).

.SH FILES
.TP
~/.gnunet/data/pseudonyms/
Directory where the pseudonyms are stored
.TP
~/.gnunet/state.sdb/collection
File where information about the currently active collection is kept (if any)
.SH "REPORTING BUGS"
Report bugs by using mantis <http://ovmj.org/~mantis/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-insert\fP(1), \fBgnunet\-search\fP(1), \fBgnunetd\fP(1)


ChangeLog
Mon Feb 28 00:33:03 EST 2005
	Releasing GNUnet 0.6.6b.

Sun Feb 27 19:27:44 EST 2005
	Fixing mantis 839.

Sun Feb 27 18:44:11 EST 2005
	Fixing mantis 833/834.

Sun Feb 27 18:15:02 EST 2005
	Fixing serious problem with LE in gnunet-insert.
	Fixing mantis 828.

Thu Feb 17 01:14:30 EST 2005
	Various memory leaks for Win32 fixed.
	Added support for GNUNET(D)_CONFIG environment variable.

Wed Feb  9 13:33:38 EST 2005
        Fixed problems with the logrotate code.  Releasing GNUnet 0.6.6a.

Sun Jan 11 19:58:13 CET 2005
	Releasing GNUnet 0.6.6

Sun Jan 11 19:24:27 CET 2005
        gnunet-gtk now resumes aborted downloads automatically.

Sun Nov 14 16:49:02 EST 2004
	Debugged collections, seems to work (not that it looks pretty,
	but it basically works, the rest is UI work).  Fixed deadlock
	in gnunet-gtk logging.  Releasing GNUnet 0.6.5.

Thu Nov  4 09:04:48 EST 2004
	Added (highly experimental) support for collections.

Sat Sep 25 22:29:48 EST 2004
	Releasing GNUnet 0.6.4a.

Sun Sep 19 23:39:04 EST 2004
	Various code cleanups and minor bugfixes.

Sun Aug 29 15:11:28 IST 2004
	Releasing GNUnet 0.6.4.

Tue Aug 24 20:59:14 IST 2004
	Significant enhancements to namespaces.  A namespace can
	now be annotated with meta-data at the time that it is
	created.
	
Sat Aug 21 01:00:06 IST 2004
	Adding GNU gettext support to configure.
	
Thu Aug 19 01:26:55 IST 2004
	Migrating gnunet-gtk to GTK 2.0 (tested with 2.4).

Mon Aug 16 00:30:59 IST 2004
	Removed gcry.  Added dependency on GNU MP library (libgmp)
	for kblocks.

Tue Aug  3 17:57:43 IST 2004
	Fixed bugs in gnunet-update.  Fixed segfault in AFS startup.
	Fixed warnings in gnunet-peer-info/gnunetd.  Fixed packaging
	error (not all of util/gcry included in distribution).
	Releasing GNUnet 0.6.3a.

Mon Aug  2 20:06:07 IST 2004
	Releasing GNUnet 0.6.3.

Sun Aug  1 01:31:19 IST 2004
	Changed most places from HexName to the new, shorter EncName.
	Pushed URIs for namespaces and keyword searches into more
	widespread use.  Replaced all occurences of sprintf with
	SNPRINTF (to obtain extra bounds check).
	
Sun Jul 11 13:54:01 IST 2004
	Moved testcases for libgnunet_util to util/ directory.
	Made DB testcases in afs/module/ run for each available
	database and not just the one configured in /etc/gnunet.conf.

Fri Jun 25 21:05:59 IST 2004
	Releasing GNUnet 0.6.2b (resolves various minor bugs).

Wed May  5 17:34:39 EST 2004
	Releasing GNUnet 0.6.2a (resolves libgcrypt and packageing
	issues, improved gtk-statistics, other minor bugfixes).
	
Fri Apr 30 21:36:23 EST 2004
	Fixed synchronization problem in cron (only relevant for
	parallel downloading).  Releasing GNUnet 0.6.2.

Mon Apr 26 21:13:45 EST 2004
	Added support to gnunet-check to convert file index database
	to new design.  Removed support for pre 0.6.1b bloomfilter
	conversion (would have been too messy to support both).
	Added old-version check to gnunetd startup to make updates
	more systematic in the future (not perfect, though).

Mon Apr 26 19:27:29 EST 2004
	Recursive insertion of directories with gnunet-gtk 
	now works including the progress bar.

Sun Apr 25 18:24:55 EST 2004
	Global changes to fix bug #698.  The fix is still 
	incomplete with respect to pthread.h specific structs.
	
Sun Apr 25 15:20:29 EST 2004
	Fixed compile error in IPv6 transports.

Fri Apr 23 23:38:01 EST 2004
	Added recursive insertion of directories to gnunet-gtk.
	Progress-bar does not work yet, also libextractor is
	always used (no option to turn it off).
	
Fri Apr 23 21:49:18 EST 2004
	Removed need to specify _protocol in configuration file
	(shorter is better).

Wed Apr 21 23:32:36 EST 2004
	Added code to bound priority used by clients by twice the
	average priority of requests from other peers that are 
	routed at the moment.  That prevents clients from exposing
	themselves with excessively high priorities (and also 
	avoids giving too much credit/trust needlessly).

Tue Apr 20 22:27:19 EST 2004
	Added possibility of using a link instead of making a copy
	if gnunetd and the insert-client both run on the same machine.

Tue Apr 20 00:00:58 EST 2004
	Added network transparency for indexing files (by copying
	them to a GNUnet(d) specific directory).  This breaks
	unindexing (for files indexed with previous versions) and
	forces the user to make a copy of the file.  In the near
	future we should resolve this problem by allowing the use
	of a link.

Sun Apr 18 17:24:34 EST 2004
	Adding recursive, parallel download of directories to
	gnunet-download

Sun Apr 18 01:57:32 EST 2004
	gnunet-transport-check can now be used to check
	external connectivity (with other peers).
	
Sat Apr 17 17:46:19 EST 2004
	Reduced memory consuption for AFS insertion and
	download.
	
Sat Apr 10 21:17:46 EST 2004
	Implemented statistics-plot in gnunet-gtk based on
	gnome-system-monitor code.

Fri Apr  9 22:01:51 EST 2004
	Added support for libextractor > 0.2.7.
	
Fri Apr  9  0:29:20 EST 2004
	Completed --help conversion.

Wed Mar 31 15:40:51 EST 2004
	Releasing 0.6.1d.

Tue Mar 30 22:42:13 EST 2004
	Integrated gnunet-pseudonym-create, -delete and -list into
	gnunet-pseudonym as well as gnunet-directory-list, -emptydb
	and -print into gnunet-directory.

Tue Mar  2 10:46:01 EST 2004
	Fixing bug preventing download for indexed files <= 1k.

Sat Feb 28 03:43:34 EST 2004
	Releasing 0.6.1c.

Sat Feb 21 06:16:47 EST 2004
	Modified requestmanager to improve performance at the end of the 
	download.

Thu Feb 19 00:48:33 EST 2004
	Fixed integer underflow possibly causing slow download performance.
	
Sun Feb 15 15:17:05 EST 2004
	Fixed weak key failures with libgcrypt.

Sun Feb 15 13:14:44 EST 2004
	New routing code (untested) commited.

Sun Feb 15 11:58:10 EST 2004
	Fixing problem with 64-bit size_t on alpha (#655).

Sat Jan 31 04:32:33 EST 2004
	Releasing 0.6.1b.

Fri Jan 23 00:00:09 EST 2004
	Fixed indexed content being overwritten by migrated content
	
Thu Jan 22 19:27:33 EST 2004
	Fixed integer-overflow problems in statuscalls.

src/applications/afs/tools/gnunet-pseudonym.c
/*
     This file is part of GNUnet.
     (C) 2002, 2003, 2004 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/afs/tools/gnunet-pseudonym.c
 * @brief create, list or delete pseudoynms
 * @author Christian Grothoff
 */

#include "gnunet_afs_esed2.h"
#include "platform.h"

static void printhelp() {
  static Help help[] = {
    { 'a', "automate", NULL,
      gettext_noop("automate creation of a namespace by starting a collection") },
    HELP_CONFIG,
    { 'C', "create", "NICKNAME",
      gettext_noop("create a new pseudonym under the given NICKNAME (with the given password if specified)") },
    { 'D', "delete", "NICKNAME",
      gettext_noop("delete the pseudonym with the given NICKNAME") },
    { 'e', "email", "EMAIL",
      gettext_noop("specify the given EMAIL address as the contact address for the pseudonym (use when creating a new pseudonym)") },
    { 'E', "end", NULL,
      gettext_noop("end automated building of a namespace (ends collection)") },
    HELP_HELP,
    HELP_LOGLEVEL,
    { 'k', "keyword", "KEYWORD",
      gettext_noop("use the given keyword to advertise the namespace (use when creating a new pseudonym)") },
    { 'm', "mimetype", "MIMETYPE",
      gettext_noop("specify that the contents of the namespace are of the given MIMETYPE (use when creating a new pseudonym)") },
    { 'n', "no-advertisement", NULL,
      gettext_noop("do not generate an advertisement for this namespace (use when creating a new pseudonym)") },
    { 'p', "password", "PASS",
      gettext_noop("use the given password to encrypt or decrypt pseudonyms in the pseudonym database") },
    { 'q', "quiet", NULL,
      gettext_noop("do not list the pseudonyms from the pseudonym database") },
    { 'r', "realname", "NAME",
      gettext_noop("specify NAME to be the realname of the user controlling the namespace (use when creating a new pseudonym)") },
    { 'R', "root", "IDENTIFIER",
      gettext_noop("specify IDENTIFIER to be the address of the entrypoint to content in the namespace (use when creating a new pseudonym)") },
    { 's', "set-rating", "ID:VALUE",
      gettext_noop("change the rating for the given namespace") },
    { 't', "text", "DESCRIPTION",
      gettext_noop("use DESCRIPTION to describe the content of the namespace (use when creating a new pseudonym)") },
    { 'u', "uri", "URI",
      gettext_noop("specify the given URI as an address that contains more information about the namespace (use when creating a new pseudonym)") },
    HELP_VERSION,
    HELP_END,
  };
  formatHelp("gnunet-pseudonym [OPTIONS]",
	     _("Create new pseudonyms, delete pseudonyms or list existing pseudonyms."),
	     help);
}

/**
 * Perform option parsing from the command line. 
 */
static int parser(int argc, 
	   char * argv[]) {
  int c;

  while (1) {
    int option_index = 0;
    static struct GNoption long_options[] = {
      LONG_DEFAULT_OPTIONS,
      { "automate", 0, 0, 'a' },
      { "create", 1, 0, 'C' },
      { "delete", 1, 0, 'D' },
      { "email", 1, 0, 'e'  },      
      { "end", 0, 0, 'E' },
      { "keyword", 1, 0, 'k' },
      { "mimetype", 1, 0, 'm' },     
      { "no-advertisement", 0, 0, 'n' },
      { "password", 1, 0, 'p' },
      { "quiet", 0, 0, 'q' },
      { "realname", 1, 0, 'r' }, 
      { "root", 1, 0, 'R' },
      { "set-rating", 1, 0, 's' },
      { "text", 1, 0, 't' },
      { "uri", 1, 0, 'u' },
      { 0,0,0,0 }
    };
    
    c = GNgetopt_long(argc,
		      argv, 
		      "ac:C:D:e:Ehk:L:m:np:qr:R:s:t:u:v", 
		      long_options, 
		      &option_index);
    
    if (c == -1) 
      break;  /* No more flags to process */
    if (YES == parseDefaultOptions(c, GNoptarg))
      continue;
    switch(c) {
    case 'a':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "AUTOMATE",
					 "START"));
      break;
    case 'C':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "CREATE",
					 GNoptarg));
      break;
    case 'D':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "DELETE",
					 GNoptarg));
      break;
    case 'e':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "EMAIL",
					 GNoptarg));
      break;
    case 'E':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "AUTOMATE",
					 "STOP"));
      break;
    case 'k':
      /* TODO: support using -k multiple times! */
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "KEYWORD",
					 GNoptarg));
      break;
    case 'h': 
      printhelp();
      return SYSERR;
    case 'm':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "MIMETYPE",
					 GNoptarg));
      break;
    case 'n':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "NO-ADVERTISEMENT",
					 "YES"));
      break;      
    case 'p':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "PASSWORD",
					 GNoptarg));
      break;
    case 'q':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "QUIET",
					 "YES"));
      break;
    case 'r':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "REALNAME",
					 GNoptarg));
      break;
    case 'R': {
      EncName enc;
      HashCode160 hc;

      if (SYSERR == enc2hash(GNoptarg,
			     &hc))
	hash(GNoptarg,
	     strlen(GNoptarg),
	     &hc);
      hash2enc(&hc, &enc);
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "ROOT",
					 (char*)&enc));
      break;
    }
    case 's':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "SET-RATING",
					 GNoptarg));
      break;
    case 't':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "DESCRIPTION",
					 GNoptarg));
      break;
    case 'u':
      FREENONNULL(setConfigurationString("PSEUDONYM",
					 "URI",
					 GNoptarg));
      break;
    case 'v':
      printf("gnunet-pseudoynm v%s\n",
	     VERSION);
      return SYSERR;
    default:
      LOG(LOG_FAILURE, 
	  _("Use --help to get a list of options.\n"));
      return SYSERR;
    } /* end of parsing commandline */
  }
  if (GNoptind < argc) {
    while (GNoptind < argc)
      LOG(LOG_WARNING, 
	  _("Invalid argument: '%s'\n"), argv[GNoptind++]);
    LOG(LOG_FATAL,
	_("Invalid arguments. Exiting.\n"));
    return SYSERR;
  }
  return OK;
}


int main(int argc, char *argv[]) {
  char ** list;
  int i;
  int cnt;
  int cpos;
  char * pass;
  char * pname;
  int success;
  Hostkey hk;
  NBlock * info;
  HashCode160 k;
  NBlock out;
	
  success = 0; /* no errors */
  if (OK != initUtil(argc, argv, &parser))
    return SYSERR;

  if (testConfigurationString("PSEUDONYM",
			      "AUTOMATE",
			      "STOP")) {
    printf(_("Collection stopped.\n"));
    stopCollection();
  }


  pname = getConfigurationString("PSEUDONYM",
				 "DELETE");
  if (pname != NULL) {
    if (OK == deletePseudonym(pname)) {
      printf(_("Pseudonym '%s' deleted.\n"),
	     pname);
    } else {
      success += 2;
      printf(_("Error deleting pseudonym '%s' (does not exist?).\n"),
	     pname);
    }
    FREE(pname);
  }

  pass = getConfigurationString("PSEUDONYM",
				"PASSWORD");
  pname = getConfigurationString("PSEUDONYM",
				 "CREATE");
  if (pname != NULL) {
    if (testConfigurationString("PSEUDONYM",
				"AUTOMATE",
				"START")) {
      char * nickname;
      char * description;
      char * realname;
      char * uri;
      char * contact;

      printf(_("Starting collection.\n"));
      hk = NULL;
      nickname = STRDUP(pname);
      description = getConfigurationString("PSEUDONYM",
					   "DESCRIPTION");
      realname = getConfigurationString("PSEUDONYM",
					"REALNAME");
      uri = getConfigurationString("PSEUDONYM",
				   "URI");
      contact = getConfigurationString("PSEUDONYM",
				       "EMAIL");
      startCollection(nickname,
		      description,
		      realname,
		      uri,
		      contact);
      FREENONNULL(nickname);
      FREENONNULL(description);
      FREENONNULL(realname);
      FREENONNULL(uri);
      FREENONNULL(contact);      
    } else {
      if ( (pass == NULL) || (pass[0]=='\n') )
	LOG(LOG_WARNING, 
	    _("No password supplied.\n"));
      
      hk = createPseudonym(pname,
			   pass);
      if (hk == NULL) {
	printf(_("Could not create pseudonym '%s' (exists?).\n"),
	       pname);
	success += 1;
      } else {
	printf(_("Pseudonym '%s' created.\n"),
	       pname);
      }
    }
    
    if ( (hk != NULL) &&
	 (! testConfigurationString("PSEUDONYM",
				    "NO-ADVERTISEMENT",
				    "YES")) ) {
      NBlock * nblock;
      char * nickname;
      char * description;
      char * realname;
      char * mimetype;
      char * uri;
      char * contact;
      char * root;
      char * summary;
      char * keyword;
      HashCode160 rootEntry;
      GNUNET_TCP_SOCKET * sock;
      
      nickname = getConfigurationString("PSEUDONYM",
					"CREATE");
      description = getConfigurationString("PSEUDONYM",
					   "DESCRIPTION");
      realname = getConfigurationString("PSEUDONYM",
					"REALNAME");
      mimetype = getConfigurationString("PSEUDONYM",
					"MIMETYPE");
      uri = getConfigurationString("PSEUDONYM",
				   "URI");
      contact = getConfigurationString("PSEUDONYM",
				       "EMAIL");
      root = getConfigurationString("PSEUDONYM",
				    "ROOT");
      if (root == NULL)
	memset(&rootEntry, 0, sizeof(HashCode160));
      else
	enc2hash(root, &rootEntry);
      
      nblock = buildNBlock(hk,
			   nickname,
			   description,
			   realname,
			   mimetype,
			   uri,
			   contact,
			   &rootEntry);
      FREENONNULL(nickname);
      FREENONNULL(description);
      FREENONNULL(realname);
      FREENONNULL(mimetype);
      FREENONNULL(uri);
      FREENONNULL(contact);
      FREENONNULL(root);	
      GNUNET_ASSERT(nblock != NULL); /* sign failed!? */
      
      decryptNBlock(nblock);
      addNamespace(nblock);
      
      summary = rootNodeToString((const RootNode*) nblock);
      printf(_("Advertising namespace with description:\n%s\n"),
	     summary);
      FREE(summary);
      
      /* now what? 
	 a) addNamespaceMetadata(nblock) [ store locally ]
	 b) encrypt with key and send to gnunetd!
      */
      sock = getClientSocket();
      if (sock == NULL)
	errexit(_("Could not connect to gnunetd.\n"));
      
      keyword = getConfigurationString("PSEUDONYM",
				       "KEYWORD");
      if (keyword == NULL)
	keyword = STRDUP("namespace"); /* default keyword */
      if (OK != insertRootWithKeyword(sock,
				      (const RootNode*) nblock,
				      keyword,
				      getConfigurationInt("GNUNET-INSERT",
							  "CONTENT-PRIORITY")))
	printf(_("Error inserting NBlock under keyword '%s'. "
		 "Is gnunetd running and space available?\n"),
	       keyword);
      FREE(keyword);	
      
      /* also publish nblock as SBlock in namespace! */
      
      
      memset(&k,0, sizeof(HashCode160));	
      encryptSBlock(&k, 
		    (const SBlock*) nblock,
		    (SBlock*) &out);
      if (OK != insertSBlock(sock,
			     (const SBlock *) &out)) 
	printf(_("Error inserting NBlock into namespace. "
		 "Is gnunetd running and space available?\n"));
      releaseClientSocket(sock);
      FREE(nblock);
    }
    if (hk != NULL)
      freeHostkey(hk);  
    FREE(pname);
  }

  if (testConfigurationString("PSEUDONYM",
			      "QUIET",
			      "YES"))
    return success; /* do not print! */

  /* print information about local pseudonyms */
  list = NULL;
  cnt = listPseudonyms(&list);
  if (cnt == -1) 
    printf(_("Could not access pseudonym directory.\n"));  
  for (i=0;i<cnt;i++) {
    const char * id;
    EncName enc;
    
    Hostkey p = readPseudonym(list[i],
			      pass);
    if (p != NULL) {
      PublicKey pk;
      HashCode160 hc;
      getPublicKey(p, &pk);
      hash(&pk, sizeof(PublicKey), &hc);
      hash2enc(&hc, &enc);
      id = (char*)&enc;
    } else
      id = _("not decrypted");
    printf(_("Pseudonym with nickname '%s' has ID '%s'.\n"),
	   list[i],
	   id);
    FREE(list[i]);
  }
  FREENONNULL(list);

  /* now print information about namespaces (local and remote) */
  info = NULL;
  cnt = listNamespaces(&info);
  if (cnt == -1) 
    printf(_("Did not find any meta-information about namespaces.\n"));  

  pname = getConfigurationString("PSEUDONYM",
				 "SET-RATING");
  if (pname != NULL) {
    if (NULL == strstr(pname, ":")) {
      fprintf(stderr,
	      _("Invalid argument '%s' for option '%s', needs a '%s'.\n"),
	      pname, "-s", ":");
      FREE(pname);
      pname = NULL;
    } else {
      cpos = (strstr(pname, ":") - pname) + 1;
    }
  }
  for (i=0;i<cnt;i++) {
    int delta;

    printNBlock(stdout, &info[i]);  
    delta = 0;
    if (pname != NULL) {
      char * iname = getUniqueNickname(&info[i].namespace);
      if ( (strlen(iname)+1 == cpos) &&
	   (0 == strncmp(pname,
			 iname,
			 cnt)) ) {
	delta = strtol(&pname[cpos],
		       NULL, /* no error handling yet */
		       10);	      
      }
      FREE(iname);
    }
    if (delta != 0) {
      fprintf(stdout, 
	      _("\tRating (before): %d\n"), 
	      evaluateNamespace(&info[i].namespace,
				0));
      evaluateNamespace(&info[i].namespace,
			delta);
      fprintf(stdout, 
	      _("\tRating (after): %d\n"), 
	      evaluateNamespace(&info[i].namespace,
				0));
    } else {
      fprintf(stdout, 
	      _("\tRating: %d\n"), 
	      evaluateNamespace(&info[i].namespace,
				0));
    }
  }
  FREENONNULL(pname);
  FREENONNULL(info);

  doneUtil();
  return success;
}

/* end of gnunet-pseudonym.c */
